system_metrics_collector/src/topic_statistics_collector/received_message_age.hpp

```diff
@@ -18,15 +18,43 @@
 #include <chrono>
 #include <string>
 #include <sstream>
+#include <type_traits>
 
 #include "topic_statistics_collector.hpp"
 
-#include "message_filters/message_traits.h"
 #include "rcl/time.h"
-#include "rclcpp/clock.hpp"
 #include "rcutils/logging_macros.h"
 
 
+namespace
+{
+template<typename M, typename = void>
+struct HasHeader : public std::false_type {};
+
+template<typename M>
+struct HasHeader<M, decltype((void) M::header)>: std::true_type {};
+
+template<typename M, typename Enable = void>
+struct TimeStamp
+{
+  static uint64_t value(const M & m)
+  {
+    (void)m;
+    return 0;
+  }
+};
+
+template<typename M>
+struct TimeStamp<M, typename std::enable_if<HasHeader<M>::value>::type>
+{
+  static uint64_t value(const M & m)
+  {
+    auto stamp = m.header.stamp;
+    return stamp.nanosec;
```

Comment:
Does `stamp` not have a `sec` component that you're dropping?

Reply:
It does, but if nanoseconds is set do we need it?

Reply:
Yeah, you do. The `sec` field is not the same thing as the `nanosec` field just in a different unit. They have a `sec` field to hold the value for when the `nanosec` field rolls over.

Reply:
e.g. nanosec <= 999,999,999